tools/ctl/command/runbenchmark/runbenchmark.go

```diff
@@ -102,6 +108,70 @@ func parseArgs(cfg *Config, args []string) error {
 	return nil
 }
 
+// compileToFS builds a Go program and loads the binary directly into the provided memfs.
+func compileToFS(ctx context.Context, fs billy.Filesystem, srcPath string) error {
+	if filepath.Base(srcPath) == "main.go" {
+		return errors.New("unexpected main.go, did you mean to provide the dir?")
+	}
+	dir, err := os.MkdirTemp("", "*")
+	if err != nil {
+		return errors.Wrap(err, "creating tempdir")
+	}
+	defer os.RemoveAll(dir)
+	log.Printf("Compiling  %s...", srcPath)
+	cmd := exec.CommandContext(ctx, "go", "build", "-o", dir, "-gcflags", "-l=4", srcPath)
+	cmd.Env = append(os.Environ(), "GOOS=linux", fmt.Sprintf("GOARCH=%s", runtime.GOARCH), "CGO_ENABLED=0")
+	if output, err := cmd.CombinedOutput(); err != nil {
+		return fmt.Errorf("compilation failed:\n%s", string(output))
+	}
+	// Assume the srcPath is either the parent dir, or source file matching the binary name.
+	binName := strings.TrimSuffix(filepath.Base(srcPath), ".go")
+	tmpName := filepath.Join(dir, binName)
+	diskBin, err := os.Open(tmpName)
+	if err != nil {
+		return err
+	}
+	defer diskBin.Close()
+	memBin, err := fs.Create(binName)
+	if err != nil {
+		return err
+	}
+	defer memBin.Close()
+	if _, err := io.Copy(memBin, diskBin); err != nil {
+		return fmt.Errorf("failed to copy binary to memfs: %w", err)
+	}
+	return nil
+}
+
+// serveFS serves the Filesystem, choosing and returning an available port
+func serveFS(fs billy.Filesystem) (int, error) {
```

Comment:
i'm realizing we'll immediately also need this in TUI so we could add a ctl-level library to do this build-and-serve dance.